flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -209,6 +212,112 @@ class BoxValue : public AbstractBox, public AbstractArrayBox {
+  /// parameters must always be provided, even if they are constant and already
+  /// reflected in the address type.
+  MutableBoxValue(mlir::Value addr, mlir::ValueRange lenParameters,
+                  MutableProperties mutableProperties)
+      : AbstractBox(addr), lenParams{lenParameters.begin(),
+                                     lenParameters.end()},
+        mutableProperties{mutableProperties} {
+    // Currently only accepts fir.(ref/ptr/heap)<fir.box<type>> mlir::Value for
+    // the address. This may change if we accept
+    // fir.(ref/ptr/heap)<fir.heap<type>> for scalar without length parameters.
+    assert(verify() &&
+           "MutableBoxValue requires mem ref to fir.box<fir.[heap|ptr]<type>>");
+  }
+
+  /// Get the fir.box<type> part of the address type.
+  fir::BoxType getBoxTy() const {
+    auto type = getAddr().getType();
+    return fir::dyn_cast_ptrEleTy(type).cast<fir::BoxType>();
+  }
+  /// Return the part of the address type after memory and box types. That is
+  /// the element type, maybe wrapped in a fir.array type.
+  mlir::Type getBaseTy() const {
+    return fir::dyn_cast_ptrEleTy(getBoxTy().getEleTy());
+  }
+  /// Get the scalar type related to the described entity
+  mlir::Type getEleTy() const {
+    auto type = getBaseTy();
+    if (auto seqTy = type.dyn_cast<fir::SequenceType>())
+      return seqTy.getEleTy();
+    return type;
+  }
+
+  /// Is this a Fortran pointer ?
+  bool isPointer() const {
+    return getBoxTy().getEleTy().isa<fir::PointerType>();
+  }
+  /// Is this an allocatable ?
+  bool isAllocatable() const {
+    return getBoxTy().getEleTy().isa<fir::HeapType>();
+  }
+  /// Is the entity an array or an assumed rank ?
+  bool hasRank() const { return getBaseTy().isa<fir::SequenceType>(); }
+  /// Is this an assumed rank ?
+  bool hasAssumedRank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    return seqTy && seqTy.hasUnknownShape();
+  }
+  /// Returns the rank of the entity. Beware that zero will be returned for
+  /// both scalars and assumed rank.
+  unsigned rank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    if (seqTy)
+      return seqTy.getDimension();
+    return 0;
+  }
+  /// Is this a character entity ?
+  bool isCharacter() const { return getEleTy().isa<fir::CharacterType>(); };
+  /// Is this a derived type entity ?
+  bool isDerived() const { return getEleTy().isa<fir::RecordType>(); };
+  /// Does this entity has any non deferred length parameters ?
+  bool hasNonDeferredLenParams() const { return !lenParams.empty(); }
+  /// Return the non deferred length parameters.
+  llvm::ArrayRef<mlir::Value> nonDeferredLenParams() const { return lenParams; }
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &,
+                                       const MutableBoxValue &);
+  LLVM_DUMP_METHOD void dump() const { operator<<(llvm::errs(), *this); }
```

Comment:
Because I copied this from `ArrayBoxValue::dump` [here](https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/include/flang/Lower/Support/BoxValue.h#L134). This is how `BoxValue::dump`, `CharArrayBoxValue::dump`... seems to be implemented. Should I change them all to the suggested form or keep as it is ?

Reply:
I wonder who did that. lol

OK, that makes sense, but now I'm sort of thinking my old code looks a bit strange. If you agree, feel free to replace them all.

Reply:
I agree, I replaced the weird dumps with `llvm::errs() << *this;`.